### PR TITLE
don't run ZSTD_TrainDictionary in BlockBasedTableBuilder if there isn't compression needed

### LINKED ISSUE #12409: ZSTD_TrainDictionary runs even when the compression is set to kNoCompression for a given level
ZSTD_TrainDictionary [[link](https://github.com/facebook/rocksdb/blob/a53ed916917fe79a6c3f07313ec7a3fa85ae6dc4/table/block_based/block_based_table_builder.cc#L1894)] runs for `SSTFileWriter::Finish` even when `bottommost_compression` option is set to `kNoCompression`. This reduces throughput for `SstFileWriter::Finish`

We construct rocksdb options using ZSTD compression for levels including 2 and above. For levels 0 and 1, we set it to `kNoCompression`. We also set `zstd_max_train_bytes` to a non-zero positive value (which is applicable for levels with ZSTD compression enabled).  These options are used for the database and also passed to `SstFileWriter` for creating sst files to be later added to that database. Since the BlockBasedTableBuilder::Finish [[link](https://github.com/facebook/rocksdb/blob/a53ed916917fe79a6c3f07313ec7a3fa85ae6dc4/table/block_based/block_based_table_builder.cc#L1892)] only checks for `zstd_max_train_bytes` to be non-zero positive value, it runs `ZSTD_TrainDictionary` even when it shouldn't since `SSTFileWriter` is operating at bottommost level
 
### Expected behavior

If the bottommost_compression or compression_per_level for a level is set to `kNoCompression`, then don't run `ZSTD_TrainDictionary`

### Actual behavior

`ZSTD_TrainDictionary` is also run for level which has `kNoCompression` set

### PR DESCRIPTION
**Pull request by kwadhwa18**
fixes #12409

### Issue

ZSTD_TrainDictionary [[link](https://github.com/facebook/rocksdb/blob/a53ed916917fe79a6c3f07313ec7a3fa85ae6dc4/table/block_based/block_based_table_builder.cc#L1894)] runs for SSTFileWriter::Finish even when bottommost_compression option is set to kNoCompression. This reduces throughput for SstFileWriter::Finish

We construct rocksdb options using ZSTD compression for levels including 2 and above. For levels 0 and 1, we set it to kNoCompression. We also set zstd_max_train_bytes to a non-zero positive value (which is applicable for levels with ZSTD compression enabled). These options are used for the database and also passed to SstFileWriter for creating sst files to be later added to that database. Since the BlockBasedTableBuilder::Finish [[link](https://github.com/facebook/rocksdb/blob/a53ed916917fe79a6c3f07313ec7a3fa85ae6dc4/table/block_based/block_based_table_builder.cc#L1892)] only checks for zstd_max_train_bytes to be non-zero positive value, it runs ZSTD_TrainDictionary even when it shouldn't since SSTFileWriter is operating at bottommost level

### Fix

If compression_type is set to kNoCompression, then don't run ZSTD_TrainDictionary and dictionary building

### Testing

I see we have tests for sst file writer with compression type set/unset. Let me know if it isn't covered and I can extend

### CODE REVIEW

**Review by ajkr — CHANGES_REQUESTED** [review]

File: table/block_based/block_based_table_builder.cc [hunk 1]
```diff
@@ -480,8 +480,8 @@ struct BlockBasedTableBuilder::Rep {
         compression_ctxs(tbo.compression_opts.parallel_threads),
         verify_ctxs(tbo.compression_opts.parallel_threads),
         verify_dict(),
-        state((tbo.compression_opts.max_dict_bytes > 0) ? State::kBuffered
-                                                        : State::kUnbuffered),
+        state((tbo.compression_type != kNoCompression) ? State::kBuffered
+                                                       : State::kUnbuffered),
```
[the remark this answers is not in the document]
Comment (ajkr) [reply table/block_based/block_based_table_builder.cc:484]: Btw, I mean we should bring it back in addition to the condition you have added here, not revert

File: table/block_based/block_based_table_builder.cc [hunk 2]
```diff
@@ -480,8 +480,10 @@ struct BlockBasedTableBuilder::Rep {
         compression_ctxs(tbo.compression_opts.parallel_threads),
         verify_ctxs(tbo.compression_opts.parallel_threads),
         verify_dict(),
-        state((tbo.compression_opts.max_dict_bytes > 0) ? State::kBuffered
-                                                        : State::kUnbuffered),
+        state((tbo.compression_opts.max_dict_bytes > 0
+               || tbo.compression_type != kNoCompression)
```
[the remark this answers is not in the document]
Comment (kwadhwa18) [reply table/block_based/block_based_table_builder.cc:484]: I was confused by the previous comment regarding the test failure

File: db/db_block_cache_test.cc [hunk 3]
```diff
@@ -1845,10 +1845,7 @@ TEST_P(DBBlockCachePinningTest, TwoLevelDB) {
   const int kNumKeysPerFile = kBlockSize * kNumBlocksPerFile / kKeySize;
 
   Options options = CurrentOptions();
-  // `kNoCompression` makes the unit test more portable. But it relies on the
-  // current behavior of persisting/accessing dictionary even when there's no
-  // (de)compression happening, which seems fairly likely to change over time.
-  options.compression = kNoCompression;
+  options.compression = kLZ4Compression;
```
Comment (ajkr) [inline db/db_block_cache_test.cc:1848]: This test needs to be conditionally compiled now. Example: https://github.com/facebook/rocksdb/blob/fc40165614e3aecc1475d61f895f508b28247c0d/utilities/blob_db/blob_db_test.cc#L634-L725

**Review by ajkr — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**